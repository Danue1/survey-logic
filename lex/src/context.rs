pub struct Context<'lex> {
    source: &'lex str,
    index: usize,
}

impl<'lex> Context<'lex> {
    pub const fn new(source: &'lex str) -> Self {
        Self { source, index: 0 }
    }

    pub const fn index(&self) -> usize {
        self.index
    }

    pub fn peek(&self) -> Option<char> {
        self.source.get(self.index..).and_then(|s| s.chars().next())
    }

    pub fn bump(&mut self) -> Option<char> {
        let c = self.peek()?;
        self.advance();
        Some(c)
    }

    #[inline]
    pub fn advance(&mut self) {
        self.index += 1;
    }

    #[inline]
    pub fn slice(&self, start: usize) -> String {
        self.source[start..self.index].to_owned()
    }
}
