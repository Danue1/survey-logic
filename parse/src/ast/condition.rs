use crate::context::{Context, State};
use syntax_kind::SyntaxKind;

impl Context {
    pub fn logical_condition(&mut self) -> State {
        let checkpoint = self.checkpoint();
        if self.atomic_condition() == State::Continue {
            return State::Continue;
        }

        while !self.is_empty() {
            self.skip_whitespace();
            if self.logical_operator() == State::Continue {
                break;
            }

            self.skip_whitespace();
            self.start_node_at(checkpoint, SyntaxKind::LOGICAL_CONDITION_NODE);
            self.atomic_condition();
            self.finish_node();
        }

        State::Stop
    }

    pub fn atomic_condition(&mut self) -> State {
        one_of!(self.group_condition(), self.binary_condition(),)
    }

    pub fn group_condition(&mut self) -> State {
        if self.peek() != SyntaxKind::LEFT_PAREN {
            return State::Continue;
        }

        self.start_node(SyntaxKind::GROUP_CONDITION_NODE);
        self.bump();
        self.skip_whitespace();
        if self.logical_condition() != State::Continue {
            self.skip_whitespace();
        }
        if self.peek() == SyntaxKind::RIGHT_PAREN {
            self.bump();
        }

        self.finish_node();
        State::Stop
    }

    pub fn logical_operator(&mut self) -> State {
        match self.peek_str().to_lowercase().as_str() {
            "and" => {
                self.start_node(SyntaxKind::AND_OPERATOR_NODE);
                self.bump();
                self.finish_node();
                State::Stop
            }
            "or" => {
                self.start_node(SyntaxKind::OR_OPERATOR_NODE);
                self.bump();
                self.finish_node();
                State::Stop
            }
            _ => State::Continue,
        }
    }

    pub fn binary_condition(&mut self) -> State {
        let checkpoint = self.checkpoint();
        if self.question_parameter() == State::Continue {
            return State::Continue;
        }

        self.start_node_at(checkpoint, SyntaxKind::COMPARISON_CONDITION_NODE);
        self.skip_whitespace();

        if let Some(kind) = self.punctuation() {
            self.skip_whitespace();
            match kind {
                PunctuationKind::Comparison => {
                    self.start_node_at(checkpoint, SyntaxKind::BINARY_COMPARISON_NODE);
                    self.singular_literal();
                    self.finish_node();
                }
                PunctuationKind::In => {
                    self.start_node_at(checkpoint, SyntaxKind::LIST_COMPARISON_NODE);
                    self.list_literal();
                    self.finish_node();
                }
                PunctuationKind::Date => {
                    self.start_node_at(checkpoint, SyntaxKind::DATE_COMPARISON_NODE);
                    self.date_literal();
                    self.finish_node();
                }
            };
        }

        self.finish_node();
        State::Stop
    }

    pub fn punctuation(&mut self) -> Option<PunctuationKind> {
        match self.peek_str().to_lowercase().as_str() {
            "=" => {
                self.start_node(SyntaxKind::EQUAL_OPERATOR_NODE);
                self.bump();
                self.finish_node();
                return Some(PunctuationKind::Comparison);
            }
            "!" => {
                let checkpoint = self.checkpoint();
                self.bump();
                if self.peek_str() == "=" {
                    self.start_node_at(checkpoint, SyntaxKind::NOT_EQUAL_OPERATOR_NODE);
                    self.bump();
                    self.finish_node();
                    return Some(PunctuationKind::Comparison);
                }

                self.start_node_at(checkpoint, SyntaxKind::ERROR);
                self.finish_node();
                return None;
            }
            comparison @ ("<" | ">") => {
                let checkpoint = self.checkpoint();
                self.bump();
                if self.peek_str() == "=" {
                    if comparison == "<" {
                        self.start_node_at(checkpoint, SyntaxKind::LESS_THAN_OR_EQUAL_OPERATOR_NODE)
                    } else {
                        self.start_node_at(
                            checkpoint,
                            SyntaxKind::GREATER_THAN_OR_EQUAL_OPERATOR_NODE,
                        )
                    };
                    self.bump();
                    self.finish_node();
                    return Some(PunctuationKind::Comparison);
                }

                if comparison == "<" {
                    self.start_node_at(checkpoint, SyntaxKind::LESS_THAN_OPERATOR_NODE);
                } else {
                    self.start_node_at(checkpoint, SyntaxKind::GREATER_THAN_OPERATOR_NODE);
                };
                self.finish_node();
                return Some(PunctuationKind::Comparison);
            }
            "in" => {
                self.start_node(SyntaxKind::IN_OPERATOR_NODE);
                self.bump();
                self.finish_node();
                return Some(PunctuationKind::In);
            }
            "after" => {
                self.start_node(SyntaxKind::AFTER_OPERATOR_NODE);
                self.bump();
                self.finish_node();
                return Some(PunctuationKind::Date);
            }
            "before" => {
                self.start_node(SyntaxKind::BEFORE_OPERATOR_NODE);
                self.bump();
                self.finish_node();
                return Some(PunctuationKind::Date);
            }
            _ => None,
        }
    }
}

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub enum PunctuationKind {
    Comparison,
    In,
    Date,
}

#[cfg(test)]
mod tests {
    macro_rules! assert_snapshot {
        ($parse:ident($source:expr)) => {
            let tokens = lex::lex($source);
            let mut context = super::Context::new(tokens);
            context.$parse();
            let node = context.finish();
            insta::assert_debug_snapshot!(node);
        };
    }

    #[test]
    fn comparison() {
        assert_snapshot!(binary_condition("Q123 = 1"));
        assert_snapshot!(binary_condition("Q123 != 1"));
        assert_snapshot!(binary_condition("Q123 < 1"));
        assert_snapshot!(binary_condition("Q123 > 1"));
        assert_snapshot!(binary_condition("Q123 <= 1"));
        assert_snapshot!(binary_condition("Q123 >= 1"));
    }

    #[test]
    fn group() {
        assert_snapshot!(group_condition("(Q123 = 1)"));
        assert_snapshot!(group_condition("((Q123 = 1))"));
    }

    #[test]
    fn logical() {
        assert_snapshot!(logical_condition("Q123 = 1 and Q456 = 2"));
        assert_snapshot!(logical_condition("Q123 = 1 and Q456 = 2 or Q789 = 3"));
    }

    #[test]
    fn complex() {
        assert_snapshot!(logical_condition("(Q123 = 1) and Q456 = 2"));
        assert_snapshot!(logical_condition("Q123 = 1 and (Q456 = 2)"));
        assert_snapshot!(logical_condition("(Q123 = 1) and (Q456 = 2)"));
        assert_snapshot!(logical_condition("(Q123 = 1 and Q456 = 2)"));
        assert_snapshot!(logical_condition("(Q123 = 1 and Q456 = 2) or Q789 = 3"));
        assert_snapshot!(logical_condition("Q123 = 1 and (Q456 = 2 or Q789 = 3)"));
    }
}
