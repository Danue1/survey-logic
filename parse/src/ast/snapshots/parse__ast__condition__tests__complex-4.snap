---
source: parse/src/ast/condition.rs
expression: node
---
GROUP_CONDITION_NODE@0..23
  LEFT_PAREN@0..1 "("
  LOGICAL_CONDITION_NODE@1..22
    COMPARISON_CONDITION_NODE@1..9
      BINARY_COMPARISON_NODE@1..9
        QUESTION_PARAMETER_NODE@1..5
          IDENTIFIER@1..2 "Q"
          NUMERIC_LITERAL_NODE@2..5
            NUMERIC@2..5 "123"
        WHITESPACE@5..6 " "
        EQUAL_OPERATOR_NODE@6..7
          EQUAL@6..7 "="
        WHITESPACE@7..8 " "
        NUMERIC_LITERAL_NODE@8..9
          NUMERIC@8..9 "1"
    WHITESPACE@9..10 " "
    AND_OPERATOR_NODE@10..13
      IDENTIFIER@10..13 "and"
    WHITESPACE@13..14 " "
    COMPARISON_CONDITION_NODE@14..22
      BINARY_COMPARISON_NODE@14..22
        QUESTION_PARAMETER_NODE@14..18
          IDENTIFIER@14..15 "Q"
          NUMERIC_LITERAL_NODE@15..18
            NUMERIC@15..18 "456"
        WHITESPACE@18..19 " "
        EQUAL_OPERATOR_NODE@19..20
          EQUAL@19..20 "="
        WHITESPACE@20..21 " "
        NUMERIC_LITERAL_NODE@21..22
          NUMERIC@21..22 "2"
  RIGHT_PAREN@22..23 ")"

